Fix the incorrect scheduling time for the first run of dag

wanlce opened the pull request:
When catchup_by_default is set to false and the start_date in the dag is the previous day, the first dispatch time of this dag is incorrect



dag setting 

    schedule_interval='30 21 * * *',
    start_date=dates.days_ago(1),

![2022-01-21_18-31](https://user-images.githubusercontent.com/55907021/150512219-b78b98cd-69b6-415d-ac5d-986b6916956e.png)

---
error
![2022-01-21_17-27](https://user-images.githubusercontent.com/55907021/150510749-fa133559-91bb-495c-8834-d46f302be5af.png)

Correct
![2022-01-21_17-30](https://user-images.githubusercontent.com/55907021/150511023-85fcdb50-3687-4d3e-be9c-285713128519.png)


<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

uranusjr commented on airflow/timetables/interval.py:
```diff
@@ -232,7 +232,7 @@ def _skip_to_latest(self, earliest: Optional[DateTime]) -> DateTime:
         if earliest is None:
             return new_start
-        return max(new_start, earliest)
+        return max(new_start, self._get_next(earliest))
```
This needs a test to ensure if `earliest` lies _exactly_ on the schedule, the first run is not one interval late. My intuition is this should use `self._align` instead, but a test would be required to verify that.

wanlce replied:
should use anlign.  :cry:

uranusjr commented on tests/timetables/test_interval_timetable.py:
```diff
@@ -40,6 +40,31 @@
+    [
+        pytest.param(CRON_TIMETABLE, id="cron"),
+    ],
+)
```
I don’t think there’s need to parametrize this. This problem can only happen for the cron timetable, not anything else.

uranusjr commented on tests/timetables/test_interval_timetable.py:
```diff
@@ -40,6 +40,31 @@
+    """If ``catchup=False`` and start_date is a day before"""
+    next_info = timetable.next_dagrun_info(
+        last_automated_data_interval=last_automated_data_interval,
+        restriction=TimeRestriction(earliest=CURRENT_TIME-datetime.timedelta(days=1), latest=None, catchup=False),
```
But can you add a vector to paramtrize `earliest`? One for this value you use here, another for something like `CURRENT_TIME - datetime.timedelta(days=1, seconds=2)`.

wanlce replied:
I am new to this, thank you for your guidance

uranusjr reviewed (changes requested):
Logic looks good to me, but the tests need some improvements. Also need to fix linter errors.

Commit:
Fix the incorrect scheduling time for the first run of dag

When Catchup_by_default is set to false and start_date in the DAG is the
previous day, the first schedule time for this DAG may be incorrect

uranusjr commented on tests/timetables/test_interval_timetable.py:
```diff
@@ -35,11 +35,32 @@
 HOURLY_RELATIVEDELTA_TIMETABLE = DeltaDataIntervalTimetable(dateutil.relativedelta.relativedelta(hours=1))
 
+CRON_TIMETABLE = CronDataIntervalTimetable("30 16 * * *", TIMEZONE)
+CRON_INTERVAL = datetime.timedelta(minutes=30, hours=16)
```
This `CRON_INTERVAL` name is confusing; the “interval” of the timetable is actually 24 hours (because jobs run everyday at 16:30). Let’s change it (although I’m not sure what name is appropriate).

wanlce replied:
Yep, you are right. How about using INTERVAL_FROM_ZERO？:smiley:

uranusjr replied:
Maybe `DELTA_FROM_MIDNIGHT` (zero sounds wrong)?

wanlce replied:
Oh, Great!  Use it
